### PR TITLE
Automation Test Framework

### PR DESCRIPTION
**Pull request by hariharan-avs**
Initial Modularized Automation Test Framework scripts created using Pytest

### CODE REVIEW

**Comment by danintel** [comment]
You need to fix your commit--it is missing the "DCO" line required by The Linux Foundation and Hyperledger. See https://github.com/hyperledger-labs/trusted-compute-framework/blob/master/CONTRIBUTING.md

For example, for you it would be something like:

File: test_automation/conftest.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    return worker_obj, sig_obj, uri_client, private_key, err_cd
+
+def initialize_objects(config, args):
+    """ Function to initialize common objects for tests. """
+
+    logger.info('***************** INTEL TRUSTED COMPUTE FRAMEWORK (TCF)*****************')
+
+    # static server_uri and private_key of client
+    server_uri = r'http://localhost:1947'
```
Comment (danintel) [inline test_automation/conftest.py]: A Python `r''` string is only needed if you have backslashes  (`\`).  Remove the `r`.

File: test_automation/work_order/test_work_order_invalid_parameter.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    # retrieve values from conftest session fixture
+    worker_obj = setup_config[0]
+    sig_obj = setup_config[1]
+    uri_client = setup_config[2]
+    private_key = setup_config[3]
+    err_cd = setup_config[4]
+
+    # input and output names
+    input_json_file = r'./work_order/workorder_input/wo_inv_param.json'
```
Comment (danintel) [inline test_automation/work_order/test_work_order_invalid_parameter.py]: A Python `r''` string is only needed if you have backslashes (`\`).  Remove the `r`.

File: test_automation/work_order/test_work_order_success.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    # retrieve values from conftest session fixture
+    worker_obj = setup_config[0]
+    sig_obj = setup_config[1]
+    uri_client = setup_config[2]
+    private_key = setup_config[3]
+    err_cd = setup_config[4]
+
+    # input and output names
+    input_json_file = r'./work_order/workorder_input/wo_success.json'
```
Comment (danintel) [inline test_automation/work_order/test_work_order_success.py]: A Python `r''` string is only needed if you have backslashes (`\`).  Remove the `r`.

File: test_automation/work_order/test_work_order_invalid_parameter.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    output_json_file_name = 'wo_inv_param'
+
+    # expected response
+    check_submit = {"error": {"code": 5}}
+    check_get_result = {"error": {"code": 2}}
+
+    # process work order
+    err_cd, response, worker_obj, sig_obj, encrypted_session_key = work_order_utility.process_work_order(input_json_file, output_json_file_name, worker_obj, sig_obj, uri_client, private_key, err_cd, check_submit, check_get_result)
```
Comment (danintel) [inline test_automation/work_order/test_work_order_invalid_parameter.py]: This line is too long and is hard to read.  Please limit lines to 80 characters.

File: test_automation/conftest.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        if response["result"]["totalCount"] == 0:
+            err_cd = 1
+            logger.info("ERROR: Failed at WorkerLookUp - No Workers exist to process workorder.")
+
+    if err_cd == 0:
+        # create worker retrieve request
+        input_worker_retrieve = '{"jsonrpc": "2.0", "method": "WorkerRetrieve", "id": 2, "params": {"workerId": ""}}'
+        input_json_str1 = json.dumps(json.loads(input_worker_retrieve))
+        logger.info("------------------Testing WorkerRetrieve------------------")
+        # Retrieving the worker id from the "WorkerLookUp" response and update the worker id information for the further json requests
```
Comment (danintel) [inline test_automation/conftest.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/work_order/test_work_order_success.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    output_json_file_name = 'wo_success'
+
+    # expected response
+    check_submit = {"error": {"code": 5}}
+    check_get_result = {"result": {"workOrderId": ""}}
+
+    # process work order
+    err_cd, response, worker_obj, sig_obj, encrypted_session_key = work_order_utility.process_work_order(input_json_file, output_json_file_name, worker_obj, sig_obj, uri_client, private_key, err_cd, check_submit, check_get_result)
```
Comment (danintel) [inline test_automation/work_order/test_work_order_success.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/work_order_utility.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    """ Function to sign the work order request. """
+
+    logger.info("------------------Signing WorkOrderSubmit------------------")
+
+    # create session_iv through enclave_helper
+    session_iv = enclave_helper.generate_sessioniv()
+    # create encrypted_session_key from session_iv and worker_encryption_key
+    encrypted_session_key = enclave_helper.generate_encrypted_session_key(session_iv, worker_obj.worker_encryption_key)
+    # sign work order submit request
+    input_json_str1 = sig_obj.generate_client_signature(input_json_str1, worker_obj, private_key, session_iv, encrypted_session_key)
```
Comment (danintel) [inline test_automation/work_order_utility.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/work_order_utility.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        err_cd = 1
+
+    return err_cd
+
+def create_work_order_get_result(work_order_id):
+    """ Function to create work order get result request. """
+
+    logger.info("------------------Constructing WorkOrderGetResult------------------")
+    # create work order get result request
+    input_workorder_getresult = input_data = '{"jsonrpc": "2.0","method": "WorkOrderGetResult","id": 11,"params": {"workOrderId": ""}}'
```
Comment (danintel) [inline test_automation/work_order_utility.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/work_order_utility.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                logger.info("ERROR: Failed at WorkOrderGetResult - Response received with error code.")
+                err_cd = 1
+                break
+
+        response = process_request(uri_client, input_json_str1, output_json_file_name)
+        time.sleep(3)
+
+    return response
+
+def process_work_order(input_json_file, output_json_file_name, worker_obj, sig_obj, uri_client, private_key, err_cd, check_submit, check_get_result):
```
Comment (danintel) [inline test_automation/work_order_utility.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/work_order_utility.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        # read json input file for the test case
+        logger.info("------------------Input file name: %s ---------------\n", input_json_file)
+        input_json_str1 = futils.read_json_file(input_json_file)
+        #----------------------------------------------------------------------------------
+        logger.info("------------------Testing WorkOrderSubmit------------------")
+
+        # create work order request
+        input_json_str1, work_order_id = create_work_order_request(input_json_str1, worker_obj)
+        # sign work order request
+        input_json_str1, session_iv, encrypted_session_key = sign_work_order_request(input_json_str1, worker_obj, sig_obj, private_key)
```
Comment (danintel) [inline test_automation/work_order_utility.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/work_order_utility.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            # check expected key of test case
+            key_name = list(check_get_result.keys())[0]
+            logger.info("key_name in expected result: %s  \n", key_name)
+            # validate response result or error codes
+            if key_name in response:
+                if "code" in check_get_result[key_name].keys():
+                    if "code" in response[key_name].keys():
+                        if response[key_name]["code"] == check_get_result[key_name]["code"]:
+                            err_cd = 0
+                            logger.info("SUCCESS: WorkOrderGetResult Processed Successfully and contains expected result code. First Check.")
```
Comment (danintel) [inline test_automation/work_order_utility.py]: This line is too long and is hard to read. Please limit lines to 80 characters.

File: test_automation/conftest.py (hunk 1 above)
Comment (hariharan-avs) [reply test_automation/conftest.py]: Removed the unnecessary r'' and added the value to default for command line option, to provide functionality to test scripts to use different uri that default localhost

File: test_automation/work_order/test_work_order_invalid_parameter.py (hunk 2 above)
Comment (hariharan-avs) [reply test_automation/work_order/test_work_order_invalid_parameter.py]: Removed the unnecessary raw string in assignment of file name

File: test_automation/work_order/test_work_order_success.py (hunk 3 above)
Comment (hariharan-avs) [reply test_automation/work_order/test_work_order_success.py]: Removed the unnecessary raw string in assignment of file name

File: test_automation/work_order/test_work_order_invalid_parameter.py (hunk 4 above)
Comment (hariharan-avs) [reply test_automation/work_order/test_work_order_invalid_parameter.py]: Fixed line length to below 80

File: test_automation/conftest.py (hunk 5 above)
Comment (hariharan-avs) [reply test_automation/conftest.py]: Fixed line length to below 80

File: test_automation/work_order_utility.py (hunk 7 above)
Comment (hariharan-avs) [reply test_automation/work_order_utility.py]: Fixed line length to below 80

File: test_automation/work_order_utility.py (hunk 8 above)
Comment (hariharan-avs) [reply test_automation/work_order_utility.py]: Fixed line length to below 80

File: test_automation/work_order_utility.py (hunk 9 above)
Comment (hariharan-avs) [reply test_automation/work_order_utility.py]: Fixed line length to below 80

File: test_automation/work_order_utility.py (hunk 10 above)
Comment (hariharan-avs) [reply test_automation/work_order_utility.py]: Fixed line length to below 80

File: test_automation/work_order_utility.py (hunk 11 above)
Comment (hariharan-avs) [reply test_automation/work_order_utility.py]: Fixed line length to below 80

File: test_automation/work_order_utility.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+    logger.info("----- Signing WorkOrderSubmit -----")
+
+    # create session_iv through enclave_helper
+    session_iv = enclave_helper.generate_sessioniv()
+    # create encrypted_session_key from session_iv and worker_encryption_key
+    encrypted_session_key = (enclave_helper.generate_encrypted_session_key
+    (session_iv, worker_obj.worker_encryption_key))
+    # sign work order submit request
+    input_json_str1 = sig_obj.generate_client_signature(input_json_str1,
```
[the remark this answers is not in the document]
Comment (hariharan-avs) [reply test_automation/work_order_utility.py]: Rebased and included changes in line with new generate client signature function in utility

File: test_automation/conftest.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    """ Function to initialize common objects for tests. """
+
+    logger.info('***** INTEL TRUSTED COMPUTE FRAMEWORK (TCF) *****')
+
+    # private_key of client
+    private_key = enclave_helper.generate_signing_keys()
+
+    # Initializing Signature object, Worker Object
+    sig_obj = signature.ClientSignature()
+    worker_obj = worker.WorkerDetails()
```
[the remark this answers is not in the document]
Comment (hariharan-avs) [reply test_automation/conftest.py]: Rebased and included changes in line with new worker class

File: test_automation/work_order/input/work_order_invalid_parameter.json [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        "payloadFormat": "pformat",
+        "resultUri": "resulturi",
+        "notifyUri": "notifyuri",
+        "workOrderId": "",
+        "workerId": "",
+        "workloadId": "source-eval",
+        "requesterId": "0x3456",
+        "workerEncryptionKey": "",
+        "dataEncryptionAlgorithm": "AES-GCM-256",
+        "encryptedSessionKey": "sessionkey",
```
[the remark this answers is not in the document]
Comment (hariharan-avs) [reply test_automation/work_order/input/work_order_invalid_parameter.json]: Removed hardcoded sessionkey in json file

File: test_automation/work_order/input/work_order_success.json [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        "workOrderId": "",
+        "workerId": "",
+        "workloadId": "heart-disease-eval",
+        "requesterId": "0x3456",
+        "workerEncryptionKey": "",
+        "dataEncryptionAlgorithm": "AES-GCM-256",
+        "encryptedSessionKey": "sessionkey",
+        "sessionKeyIv": "Iv",
+        "requesterNonce": "",
+        "encryptedRequestHash": "requesthash",
```
[the remark this answers is not in the document]
Comment (hariharan-avs) [reply test_automation/work_order/input/work_order_success.json]: Removed hardcoded sessionkey in json file

File: test_automation/work_order/input/work_order_invalid_parameter.json [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+        "sessionKeyIv": "Iv",
+        "requesterNonce": "",
+        "encryptedRequestHash": "requesthash",
```
[the remark this answers is not in the document]
Comment (hariharan-avs) [reply test_automation/work_order/input/work_order_invalid_parameter.json]: Removed hardcoded encryptedRequestHash in json file

**Pull request closed without merging**